### DISCUSSION TITLE
How to process many mutations in a loop without reaching connection limit?

### DISCUSSION THREAD

**Discussion by seed-of-apricot** [prisma/prisma#6080] [Q&A]
When I have something like this:

```ts
bigList.map(async (element) => {
  const someList = getSomeList(element);
  (await someList).map(
    async (item) =>
      await prisma.some.upsert({
        where: {
          id: item.id,
        },
        update: {},
        create: item, // create some relations to other models
      })
  );
});
```

I always get `Timed out fetching a new connection from the pool. Please consider reducing...` error even if I increase the connection limit. I suppose this is because there are so many `upsert()` commands running at once in the nested async loop.

How could it be addressed? I couldn't use `createMany()` because there are some relations added in the upsert command and there might be a record already existing on the database.

**Comment by ryands17** [comment]
@seed-of-apricot 👋 

In this case, it would be best to split the list into chunks of maybe 50 or 100 records and then processing them 1 chunk at a time until `createMany` supports relations: https://github.com/prisma/prisma/issues/5455

**Reply by seed-of-apricot** [reply]
Thank you @ryands17! Indeed, when I use a for loop instead of an async map it works well. It seems primitive but a good workaround till we have `createMany`.

**Reply by ryands17** [reply]
Another good library if you do not want to use a `for` loop is `p-map`: https://github.com/sindresorhus/p-map

**Reply by seed-of-apricot** [reply]
That seems good, thank you 👍

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
